### PR TITLE
fix: dependence on specific attributes in android styles

### LINKED ISSUE #521: Android style limitations
## Use case
In Beagle Android, component customizations are made from attributes defined within the framework. The problem with this is that the user must change all his predefined styles to be supported in Beagle.

## Proposal
The idea is to remove Beagle's unique attributes and start treating them as styles per widget, applying through ContextThemeWrapper.
Work started on the branch https://github.com/ZupIT/beagle/tree/fix/styles

### LINKED ISSUE #581: Image component crashes application when trying to use unavailable resource
## Description
When using the `Image` component in Beagle and referencing a resource not accessible by the device's API, the application crashes.

## Steps To Reproduce
1. Use an API 19 device
2. Add a drawable to the resources folder `drawable-21`
3. Reference this resource in the application's DesignSystem
4. Assign this resource to the `Image` component
5. The application crashes

## Expected Results
When informing an image that is not accessible by the component, the application should not crash but should report the error to the user via log.

### LINKED ISSUE #590: TextInput is disabled when the style is null on Android
## Description
When `TextInput` is used and there is no defined style for the application's DesignSystem, the component is disabled.

## Steps To Reproduce
1. Return null for input styles
2. Add a TextInput to the screen
3. The component is disabled

## Expected Results
Even without receiving any style defined by the design system, the component must behave like a standard `EditText`.

### PR DESCRIPTION
**Pull request by matheusribeirozup**
## Description
This PR removes the dependency of the styles declared in Beagle that contain the exclusive attributes of Android. Now he can use the same styles as the application's design system without changes.

## Related Issues
Fixes #521, fixes #581 and fixes #590 

## Tests

I added the following tests:
`getTabBarTypedArray_should_call_BeagleTabBarStyle`,
`getInputTextStyle_should_return_default_when_inputTextStyle_is_null` and
`getInputTextStyle_should_return_a_inputTextStyle_when_is_not_null`

## Checklist
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [DCO].
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change
Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [x] Yes, this is a breaking change. *If not, delete the remainder of this section.*

### CODE REVIEW

**Comment by botbeaglezup** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: android/beagle/src/main/java/br/com/zup/beagle/android/components/Button.kt [hunk 1]
```diff
@@ -59,14 +58,15 @@ data class Button(
     @Transient
     private val preFetchHelper: PreFetchHelper = PreFetchHelper()
 
+    @Transient
```
Comment (tuliopereirazup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/components/Button.kt:61]: One thing that we could try to start doing is moving all dependencies to a internal constructor.
Let's try with this one?

example:
```kotlin
internal constructor(
   text: Bind<String>,
   styleId: String? = null,
   onPress: List<Action>? = null,
   clickAnalyticsEvent: ClickEvent? = null,
  preFetchHelper: PreFetchHelper = PreFetchHelper(),
  buttonStyle: ..
) : super(text, styleId, onPerss, clickAnalyticsEvent) {
   
   private val preFetchHelper: PreFetchHelper = preFetchHelper
}
```

File: android/beagle/src/main/java/br/com/zup/beagle/android/components/Button.kt [hunk 2]
```diff
@@ -76,30 +76,16 @@ data class Button(
...
-                background = typeArray.getDrawable(R.styleable.BeagleButtonStyle_background)
-                isAllCaps = typeArray.getBoolean(R.styleable.BeagleButtonStyle_textAllCaps, true)
-                typeArray.recycle()
-            }
-
-            styleManagerFactory.getButtonStyle(style)?.let { buttonStyle ->
-                TextViewCompat.setTextAppearance(this, buttonStyle)
-            }
-        }
+        TextViewCompat.setTextAppearance(this, buttonStyle)
```
Comment (tuliopereirazup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/components/Button.kt:85]: Using style on constructor does works for appearance as well?

Comment (matheusribeirozup) [reply android/beagle/src/main/java/br/com/zup/beagle/android/components/Button.kt:85]: I forgot this. Thanks

File: android/beagle/src/main/java/br/com/zup/beagle/android/components/Button.kt (hunk 1 above)
Comment (matheusribeirozup) [reply android/beagle/src/main/java/br/com/zup/beagle/android/components/Button.kt:61]: The problem with this change is that the variables within the secondary constructor must already have been defined in the main class, and for this reason, it must be var or lateinit var. Another point is that this builder should also always be called.

Comment (uziasferreirazup) [reply android/beagle/src/main/java/br/com/zup/beagle/android/components/Button.kt:59]: hey @matheusribeirozup I think the suggestion they need, it is this:
```
data class Button internal constructor(
    val text: Bind<String>,
    val styleId: String? = null,
    val onPress: List<Action>? = null,
    val clickAnalyticsEvent: ClickEvent? = null,
    private val viewFactory: ViewFactory = ViewFactory(),
    private val preFetchHelper: PreFetchHelper = PreFetchHelper()
) : WidgetView() {

    constructor(
        text: String,
        styleId: String? = null,
        onPress: List<Action>? = null,
        clickAnalyticsEvent: ClickEvent? = null
    ) : this(
        valueOf(text),
        styleId,
        onPress,
        clickAnalyticsEvent
    )

    constructor(
        text: Bind<String>,
        styleId: String? = null,
        onPress: List<Action>? = null,
        clickAnalyticsEvent: ClickEvent? = null
    ) : this(
        text,
        styleId,
        onPress,
        clickAnalyticsEvent,
        ViewFactory(),
        PreFetchHelper()
    )
}
```

Comment (matheusribeirozup) [reply android/beagle/src/main/java/br/com/zup/beagle/android/components/Button.kt:59]: That makes sense

**Comment by botbeaglezup** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Tests have failed, see below for more information.</td>
    </tr>
  </tbody>
</table>

### Tests: 

Name | Classname | Time|
--- | --- | ---|
make_should_return_moshi_to_serialize_a_FormSubmit | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.169|
make_should_return_moshi_to_serialize_a_TabView | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.017|
make_should_return_moshi_to_serialize_a_FormInput | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.021|
make_should_return_moshi_to_serialize_a_ScrollView | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.017|
make_should_return_moshi_to_deserialize_a_CustomWidget | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.018|
make_should_return_moshi_to_serialize_a_UndefinedComponent_of_type_PageIndicatorComponent | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.014|
make_should_return_moshi_to_deserialize_a_PageView | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.032|
make_should_return_moshi_to_serialize_a_UndefinedComponent_of_type_InputWidget | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.017|
moshi_should_serialize_bindComponent | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.023|
make_should_return_moshi_to_serialize_a_UndefinedAction | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.017|
moshi_should_deserialize_internalObject_using_component_type_attribute | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.012|
make_should_return_moshi_to_deserialize_a_ListView | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.026|
make_should_return_moshi_to_serialize_a_PageView | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.024|
make_should_return_moshi_to_deserialize_a_TabView | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.03|
make_should_return_moshi_to_serialize_a_Image_Local | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.016|
make_should_return_moshi_to_deserialize_a_Container | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.021|
make_should_return_moshi_to_serialize_a_ListView | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.034|
make_should_return_moshi_to_deserialize_a_PageIndicator | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.016|
make_should_return_moshi_to_serialize_a_Form | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.017|
make_should_return_moshi_to_serialize_a_Text | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.021|
make_should_return_moshi_to_serialize_a_ScreenWidget | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.013|
make_should_return_moshi_to_serialize_a_Container | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.014|
moshi_should_deserialize_bindComponent | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.012|
make_should_return_moshi_to_deserialize_a_PushView_action | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.02|
make_should_return_moshi_to_deserialize_a_ScreenWidget | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.016|
make_should_return_moshi_to_serialize_a_UndefinedComponent | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.056|
make_should_return_moshi_to_serialize_a_PageIndicator | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.014|
make_should_return_moshi_to_deserialize_a_Image_Local | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.012|
make_should_return_moshi_to_deserialize_a_LazyComponent | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.012|
make_should_return_moshi_to_serialize_a_Button | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.013|
make_should_return_moshi_to_deserialize_a_UndefinedComponent | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.021|
make_should_return_moshi_to_deserialize_a_Form | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.022|
make_should_return_moshi_to_deserialize_a_Text | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.013|
make_should_return_moshi_to_serialize_a_NetworkImage | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.014|
make_should_return_moshi_to_deserialize_a_FormInput | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.012|
make_should_return_moshi_to_deserialize_a_FormSubmit | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.018|
make_should_return_moshi_to_serialize_a_LazyComponent | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.013|
make_should_return_moshi_to_deserialize_a_Button | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.014|
make_should_return_moshi_to_serialize_a_CustomWidget | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.015|
make_should_return_moshi_to_deserialize_a_ScrollView | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.014|
make_should_return_moshi_to_deserialize_a_NetworkImage | br.com.zup.beagle.android.data.serializer.BeagleMoshiTest | 0.013|


<p align="right" data-meta="generated_by_android">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Pull request merged**